### PR TITLE
Introduce explicit Java modules

### LINKED ISSUE #1091: Add module descriptors to all JUnit 5 modules
## Overview

Conceal internal packages by moving from Level 1 "`Automatic-Module-Name`" to Level 2 "Publish modularized JAR": https://vimeo.com/289846017#t=1298s `21:38`

## Draft Pull Requests

* #1593 Modules via using Java 9 as release target  _outdated_
* #1844 Modules via extra compilation task
* #1846 Modules via @moditect
* #1847 Modules via `module` source set
* #1848 Modules via `src/modules` directory using `javac`
* #1857 Modules via extra compilation and jar tasks 💯 

## Deliverables

- [x] Create many drafts
- [x] Select one draft and make it happen

1. `org.junit.platform.commons`
1. `org.junit.platform.console`
1. `org.junit.platform.engine`
1. `org.junit.platform.launcher`
1. `org.junit.platform.reporting`
1. `org.junit.platform.runner`
1. `org.junit.platform.suite.api`
1. `org.junit.platform.testkit`
1. `org.junit.jupiter`
1. `org.junit.jupiter.api`
1. `org.junit.jupiter.engine`
1. `org.junit.jupiter.migrationsupport`
1. `org.junit.jupiter.params`
1. `org.junit.vintage.engine`

- [ ] Include `module-info.class` in https://github.com/ota4j-team/opentest4j/pull/51
- [ ] Include `module-info.class` in https://github.com/apiguardian-team/apiguardian

### LINKED ISSUE #1216: Revise layout and building of multi-release JARs
## Overview

Current "solution" uses external `jar` tool:
https://github.com/junit-team/junit5/blob/master/build.gradle#L489-L519

Better Gradle intergration presented by @melix at:
https://blog.gradle.org/mrjars

## Deliverables

- [ ] Revise layout and building of multi-release-JAR(s). Keep in mind, that the `jar` tool offers some API/ABI sanity checks when it packages the version-aware types.

### PR DESCRIPTION
**Pull request by sormuras**
## Overview

This PR solves #1091 -- and supersedes all draft PRs and branches created lately.

Closes #1091 
Closes #1216 
Closes #1844 
Closes #1846 
Closes #1847 
Closes #1848  
Closes #1854 

## TODO

- [x] Create Gradle `mainRelease9` `SourceSet` in `junit-platform-commons` as a container for the 9er MR-JAR version of `ModuleUtils`.
- [x] Enumerate all required modules in each module, although they are added via `transitive` directives. Effectively sync the `requires` directives with the `dependencies` block declared in our Gradle build.
- [x] Remove internal and almost dead [`ClasspathScanningSupport`](https://github.com/junit-team/junit5/blob/master/junit-platform-engine/src/main/java/org/junit/platform/engine/support/filter/ClasspathScanningSupport.java) class.
- [x] @sbrannen please verify the project imports hasslefree into Eclipse -- we might need to exlude `mainRelease9` from the `.classpath` file

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-platform-commons/junit-platform-commons.gradle.kts [hunk 1]
```diff
@@ -1,13 +1,50 @@
...
 }
 
 dependencies {
+	add("mainRelease9Compile", sourceSets.main.get().output)
 	api("org.apiguardian:apiguardian-api:${Versions.apiGuardian}")
 }
+
+tasks {
+
+	named<JavaCompile>("compileMainRelease9Java").configure {
```
Comment (marcphilipp) [inline junit-platform-commons/junit-platform-commons.gradle.kts]:
```suggestion
	val compileMainRelease9Java by existing(JavaCompile::class) {
```

File: junit-platform-commons/junit-platform-commons.gradle.kts [hunk 2]
```diff
@@ -1,13 +1,50 @@
...
+		sourceCompatibility = "9"
+		targetCompatibility = "9"
+		options.compilerArgs.addAll(listOf("--release", "9"))
+	}
+
+	jar {
+		dependsOn("compileMainRelease9Java")
```
Comment (marcphilipp) [inline junit-platform-commons/junit-platform-commons.gradle.kts]:
```suggestion
		dependsOn(compileMainRelease9Java)
```

File: junit-platform-console/junit-platform-console.gradle.kts [hunk 3]
```diff
@@ -27,6 +25,11 @@ tasks {
 			include("LICENSE-picocli.md")
 			into("META-INF")
 		}
+		doLast {
+			ToolProvider.findFirst("jar").get().run(System.out, System.err, "--update",
+					"--file", archiveFile.get().asFile.absolutePath,
+					"--main-class", "org.junit.platform.console.ConsoleLauncher")
+		}
```
Comment (marcphilipp) [inline junit-platform-console/junit-platform-console.gradle.kts]: ❓ The `Main-Class` attribute is already set below. Why is this necessary?

File: buildSrc/src/main/kotlin/java-library-conventions.gradle.kts [hunk 4]
```diff
@@ -136,46 +141,67 @@ tasks.jar {
...
-	if (automaticModuleName != null) {
-		tasks.jar {
-			manifest {
-				attributes("Automatic-Module-Name" to automaticModuleName)
-			}
-		}
-	}
+val compileModule by tasks.registering(JavaCompile::class) {
+	destinationDir = file("$buildDir/classes/java/module")
+	source = fileTree("src/module/$javaModuleName").asFileTree
```
[the remark this answers is not in the document]
Comment (marcphilipp) [reply buildSrc/src/main/kotlin/java-library-conventions.gradle.kts]:
```suggestion
	source = fileTree("src/module/$javaModuleName")
```

File: build.gradle.kts [hunk 5]
```diff
@@ -45,6 +45,7 @@ val vintageProjects by extra(listOf(
 ))
 
 val mavenizedProjects by extra(platformProjects + jupiterProjects + vintageProjects)
+val modularProjects by extra(mavenizedProjects - listOf(project(":junit-platform-console-standalone")))
```
[the remark this answers is not in the document]
Comment (sormuras) [reply build.gradle.kts]: Good idea. Even better than declaring them in the build script files.

File: junit-platform-commons-java-9/src/test/java/org/junit/platform/commons/util/ModuleUtilsTests.java [hunk 6]
```diff
@@ -1,61 +0,0 @@
...
-import java.util.Set;
-
-import org.junit.jupiter.api.Test;
-
-/**
- * Unit tests for {@link ModuleUtils}.
- *
- * @since 1.1
- */
-class ModuleUtilsTests {
```
Comment (marcphilipp) [inline junit-platform-commons-java-9/src/test/java/org/junit/platform/commons/util/ModuleUtilsTests.java]: :x: Where did this test go?

File: junit-platform-console/junit-platform-console.gradle.kts (hunk 3 above)
Comment (sormuras) [reply junit-platform-console/junit-platform-console.gradle.kts]: The `MANIFEST.MF` entry is only evaluated when running `java` in _executable jar_ mode.

Here we're talking _launching the main class in a module_: `java [options] --module module[/mainclass] [args...]` -- for when you want to omit the `/mainclass` part.

More details at: https://docs.oracle.com/en/java/javase/11/tools/java.html#GUID-3B1CE181-CD30-4178-9602-230B800D4FAE

File: junit-platform-commons-java-9/src/test/java/org/junit/platform/commons/util/ModuleUtilsTests.java (hunk 6 above)
Comment (sormuras) [reply junit-platform-commons-java-9/src/test/java/org/junit/platform/commons/util/ModuleUtilsTests.java]: 't was lost in translation. Will restore it.

File: junit-platform-console/junit-platform-console.gradle.kts (hunk 3 above)
Comment (sormuras) [reply junit-platform-console/junit-platform-console.gradle.kts]: ...and there's **no** `mainclass` directive with the `module` specification: https://docs.oracle.com/javase/specs/jls/se11/html/jls-7.html#jls-7.7

Comment (marcphilipp) [reply junit-platform-console/junit-platform-console.gradle.kts]: Thanks for the explanation!

**Comment by marcphilipp** [comment]
I've excluded the `mainRelease9` from the Eclipse config for now.

**Comment by sormuras** [comment]
Due to the internal, yet large impact of

- [ ] Deprecate `PreconditionViolationException` in package `org.junit.platform.commons.util` in favor a new sibling of `JUnitException` in the exported package `org.junit.platform.commons.util`.

...mind its usages in JUnit 5's code base:

image

...I'll move that task into its own dedicated issue. It will be addressed after this PR is merged.
